design/baremetal-operator/raid-disk-controller.md

```diff
@@ -0,0 +1,193 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# raid-extension-for-physical-disks-and-controller
+
+## Status
+
+provisional
+
+## Summary
+
+implementation for the same.
+
+## Motivation
+
+The user should be able to specify exactly which physical disks and RAID
+controllers are to be used to construct the hardware RAID volume(s).
+
+### Goals
+
+The primary goal is to depict the physical disk and RAID controller names in
+the baremetal-host ``hardware-raid`` section.
+
+Afterwards, implementation of the same will be done within the
+``baremetal-operator`` by extending the ``baremetal-host`` specification.
+
+### Non-Goals
+
+- This specification does not deal with ``software-raid`` and extension for it.
+- It does not attempt to cover any generic (vendor agnostic)  naming convention
+  for disks or controllers.
+- It does not cover testing for hardware from all the vendors. It will only be
+  tested on Dell EMC hardware, and other vendors will have to test it on their
+hardware.
+
+## Proposal
+
+The proposal is to allow specifying the physical disks and RAID controller
+names in the hardware RAID section. For this purpose, new YAML attributes are
+to added in the BareMetalHost spec, namely `physicalDisks` and `controller`.
+
+In terms of the CRD, the current implementation of the RAID has the following
+sections in it, for hardware RAID:
+
+```yaml
+spec:
+  raid:
+    hardwareRAIDVolumes:
+      - sizeGibiBytes: 1000
+        level: 1
+        name: <volume-name>
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+We propose to add fields into the CRD, in the following manner:
+
+```yaml
+spec:
+  raid:
+    hardwareRAIDVolumes:
+      - sizeGibiBytes: 1000
+        level: 1
+        name: <volume-1-name>
+        rotational: true
+        numberOfPhysicalDisks: 2
```

Comment:
If `numberOfPhysicalDisks` didn't match the length of `physicalDisks`, Ironic would throw an error at runtime:

https://docs.openstack.org/ironic/latest/admin/raid.html#backing-physical-disks

> If properties from both “Backing physical disk hints” or “Backing physical disks” are specified, they should be consistent with each other. If they are not consistent, then the RAID configuration will fail (because the appropriate backing physical disks could not be found).

I don't think there's an easy way to validate with kubebuilder this scenario, and I don't think it needs validation, but maybe there is a way.

Reply:
Okay, shall I add a note of that here?